cmd/pebble/random.go

```diff
@@ -0,0 +1,144 @@
+// of this source code is governed by a BSD-style license that can be found in
+// the LICENSE file.
+
+package main
+
+import (
+	"crypto/sha1"
+	"encoding/binary"
+	"hash"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/exp/rand"
+)
+
+func randomBlock(
+	r *rand.Rand, minBlockBytes, maxBlockBytes int, targetCompressionRatio float64,
+) []byte {
+	size := r.Intn(maxBlockBytes-minBlockBytes+1) + minBlockBytes
+	uniqueSize := int(float64(size) / targetCompressionRatio)
+	if uniqueSize < 1 {
+		uniqueSize = 1
+	}
+	data := make([]byte, size)
+	offset := 0
+	for offset+8 <= uniqueSize {
+		binary.LittleEndian.PutUint64(data[offset:], r.Uint64())
+		offset += 8
+	}
+	word := r.Uint64()
+	for offset < uniqueSize {
+		data[offset] = byte(word)
+		word >>= 8
+		offset++
+	}
+	for offset < size {
+		data[offset] = data[offset-uniqueSize]
+		offset++
+	}
+	return data
+}
+
+type sequence struct {
+	val         int64
+	cycleLength int64
+	seed        int64
+}
+
+func (s *sequence) write() int64 {
+	return (atomic.AddInt64(&s.val, 1) - 1) % s.cycleLength
+}
+
+// read returns the last key index that has been written. Note that the returned
+// index might not actually have been written yet, so a read operation cannot
+// require that the key is present.
+func (s *sequence) read() int64 {
+	return atomic.LoadInt64(&s.val) % s.cycleLength
+}
+
+// keyGenerator generates read and write keys. Read keys may not yet exist and
+// write keys may already exist.
+type keyGenerator interface {
+	writeKey() int64
+	readKey() int64
+	rand() *rand.Rand
+	sequence() int64
+}
+
+type hashGenerator struct {
+	seq    *sequence
+	random *rand.Rand
+	hasher hash.Hash
+	buf    [sha1.Size]byte
+}
+
+func newHashGenerator(seq *sequence) *hashGenerator {
+	return &hashGenerator{
+		seq:    seq,
+		random: rand.New(rand.NewSource(uint64(time.Now().UnixNano()))),
+		hasher: sha1.New(),
+	}
+}
+
+func (g *hashGenerator) hash(v int64) int64 {
+	binary.BigEndian.PutUint64(g.buf[:8], uint64(v))
+	binary.BigEndian.PutUint64(g.buf[8:16], uint64(g.seq.seed))
+	g.hasher.Reset()
+	_, _ = g.hasher.Write(g.buf[:16])
+	g.hasher.Sum(g.buf[:0])
+	return int64(binary.BigEndian.Uint64(g.buf[:8]))
+}
+
+func (g *hashGenerator) writeKey() int64 {
+	return g.hash(g.seq.write())
+}
+
+func (g *hashGenerator) readKey() int64 {
+	v := g.seq.read()
+	if v == 0 {
+		return 0
+	}
+	return g.hash(g.random.Int63n(v))
+}
+
+func (g *hashGenerator) rand() *rand.Rand {
+	return g.random
+}
+
+func (g *hashGenerator) sequence() int64 {
+	return atomic.LoadInt64(&g.seq.val)
+}
+
+type sequentialGenerator struct {
+	seq    *sequence
+	random *rand.Rand
+}
+
+func newSequentialGenerator(seq *sequence) *sequentialGenerator {
+	return &sequentialGenerator{
+		seq:    seq,
+		random: rand.New(rand.NewSource(uint64(time.Now().UnixNano()))),
+	}
+}
+
+func (g *sequentialGenerator) writeKey() int64 {
+	return g.seq.write()
+}
+
+func (g *sequentialGenerator) readKey() int64 {
```

Comment:
We could probably change to uint64 keys now, but I tried to avoid sneaking in too many changes to keep this PR a reasonably clean copy.